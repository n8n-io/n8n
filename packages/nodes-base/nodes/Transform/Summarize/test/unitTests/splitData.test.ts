import {
	fieldValueGetter,
	aggregateAndSplitData,
	flattenAggregationResultToArray,
	type Aggregations,
} from '../../utils';

describe('Test Summarize Node, aggregateAndSplitData', () => {
	test('should not convert strings to numbers', () => {
		const data = [
			{
				Sku: '012345',
				Warehouse: 'BER_0G',
				Qty: '1',
				_itemIndex: 0,
			},
			{
				Sku: '012345',
				Warehouse: 'BER_0L',
				Qty: '2',
				_itemIndex: 1,
			},
			{
				Sku: '06534563534',
				Warehouse: 'BER_0L',
				Qty: '1',
				_itemIndex: 2,
			},
		];

		const aggregations: Aggregations = [
			{
				aggregation: 'append',
				field: 'Warehouse',
				includeEmpty: true,
			},
		];

		const result = aggregateAndSplitData({
			splitKeys: ['Sku', 'Qty'],
			inputItems: data,
			fieldsToSummarize: aggregations,
			options: { continueIfFieldNotFound: true },
			getValue: fieldValueGetter(),
		});
		expect(result).toMatchSnapshot('result');
		expect(flattenAggregationResultToArray(result)).toMatchSnapshot('array');
	});

	test('should not convert numbers to strings', () => {
		const data = [
			{
				Sku: 12345,
				Warehouse: 'BER_0G',
				Qty: 1,
				_itemIndex: 0,
			},
			{
				Sku: 12345,
				Warehouse: 'BER_0L',
				Qty: 2,
				_itemIndex: 1,
			},
			{
				Sku: 6534563534,
				Warehouse: 'BER_0L',
				Qty: 1,
				_itemIndex: 2,
			},
		];

		const aggregations: Aggregations = [
			{
				aggregation: 'append',
				field: 'Warehouse',
				includeEmpty: true,
			},
		];

		const result = aggregateAndSplitData({
			splitKeys: ['Sku', 'Qty'],
			inputItems: data,
			fieldsToSummarize: aggregations,
			options: { continueIfFieldNotFound: true },
			getValue: fieldValueGetter(),
		});
		expect(result).toMatchSnapshot('result');
		expect(flattenAggregationResultToArray(result)).toMatchSnapshot('array');
	});

	describe('with skipEmptySplitFields=true', () => {
		test('should skip empty split fields', () => {
			const data = [
				{
					Sku: 12345,
					Warehouse: 'BER_0G',
					_itemIndex: 0,
				},
				{
					Warehouse: 'BER_0L',
					_itemIndex: 1,
				},
				{
					Sku: {},
					Warehouse: 'BER_0L',
					_itemIndex: 2,
				},
				{
					Sku: 12345,
					Warehouse: { name: 'BER_0G3' },
					_itemIndex: 3,
				},
			];

			const aggregations: Aggregations = [
				{
					aggregation: 'concatenate',
					field: 'Warehouse',
					separateBy: 'other',
					customSeparator: '//',
				},
			];

			const result = aggregateAndSplitData({
				splitKeys: ['Sku'],
				inputItems: data,
				fieldsToSummarize: aggregations,
				options: { continueIfFieldNotFound: true, skipEmptySplitFields: true },
				getValue: fieldValueGetter(),
			});
			expect(result).toMatchSnapshot('result');
			expect(flattenAggregationResultToArray(result)).toMatchSnapshot('array');
		});
	});
});
