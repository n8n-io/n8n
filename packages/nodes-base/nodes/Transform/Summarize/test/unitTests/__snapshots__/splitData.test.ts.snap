// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Summarize Node, aggregateAndSplitData should not convert numbers to strings: array 1`] = `
[
  {
    "pairedItems": [
      0,
    ],
    "returnData": {
      "Qty": 1,
      "Sku": 12345,
      "appended_Warehouse": [
        "BER_0G",
      ],
    },
  },
  {
    "pairedItems": [
      1,
    ],
    "returnData": {
      "Qty": 2,
      "Sku": 12345,
      "appended_Warehouse": [
        "BER_0L",
      ],
    },
  },
  {
    "pairedItems": [
      2,
    ],
    "returnData": {
      "Qty": 1,
      "Sku": 6534563534,
      "appended_Warehouse": [
        "BER_0L",
      ],
    },
  },
]
`;

exports[`Test Summarize Node, aggregateAndSplitData should not convert numbers to strings: result 1`] = `
{
  "fieldName": "Sku",
  "splits": Map {
    12345 => {
      "fieldName": "Qty",
      "splits": Map {
        1 => {
          "pairedItems": [
            0,
          ],
          "returnData": {
            "appended_Warehouse": [
              "BER_0G",
            ],
          },
        },
        2 => {
          "pairedItems": [
            1,
          ],
          "returnData": {
            "appended_Warehouse": [
              "BER_0L",
            ],
          },
        },
      },
    },
    6534563534 => {
      "fieldName": "Qty",
      "splits": Map {
        1 => {
          "pairedItems": [
            2,
          ],
          "returnData": {
            "appended_Warehouse": [
              "BER_0L",
            ],
          },
        },
      },
    },
  },
}
`;

exports[`Test Summarize Node, aggregateAndSplitData should not convert strings to numbers: array 1`] = `
[
  {
    "pairedItems": [
      0,
    ],
    "returnData": {
      "Qty": "1",
      "Sku": "012345",
      "appended_Warehouse": [
        "BER_0G",
      ],
    },
  },
  {
    "pairedItems": [
      1,
    ],
    "returnData": {
      "Qty": "2",
      "Sku": "012345",
      "appended_Warehouse": [
        "BER_0L",
      ],
    },
  },
  {
    "pairedItems": [
      2,
    ],
    "returnData": {
      "Qty": "1",
      "Sku": "06534563534",
      "appended_Warehouse": [
        "BER_0L",
      ],
    },
  },
]
`;

exports[`Test Summarize Node, aggregateAndSplitData should not convert strings to numbers: result 1`] = `
{
  "fieldName": "Sku",
  "splits": Map {
    "012345" => {
      "fieldName": "Qty",
      "splits": Map {
        "1" => {
          "pairedItems": [
            0,
          ],
          "returnData": {
            "appended_Warehouse": [
              "BER_0G",
            ],
          },
        },
        "2" => {
          "pairedItems": [
            1,
          ],
          "returnData": {
            "appended_Warehouse": [
              "BER_0L",
            ],
          },
        },
      },
    },
    "06534563534" => {
      "fieldName": "Qty",
      "splits": Map {
        "1" => {
          "pairedItems": [
            2,
          ],
          "returnData": {
            "appended_Warehouse": [
              "BER_0L",
            ],
          },
        },
      },
    },
  },
}
`;

exports[`Test Summarize Node, aggregateAndSplitData with skipEmptySplitFields=true should skip empty split fields: array 1`] = `
[
  {
    "pairedItems": [
      0,
      3,
    ],
    "returnData": {
      "Sku": 12345,
      "concatenated_Warehouse": "BER_0G//{"name":"BER_0G3"}",
    },
  },
  {
    "pairedItems": [
      2,
    ],
    "returnData": {
      "Sku": "{}",
      "concatenated_Warehouse": "BER_0L",
    },
  },
]
`;

exports[`Test Summarize Node, aggregateAndSplitData with skipEmptySplitFields=true should skip empty split fields: result 1`] = `
{
  "fieldName": "Sku",
  "splits": Map {
    12345 => {
      "pairedItems": [
        0,
        3,
      ],
      "returnData": {
        "concatenated_Warehouse": "BER_0G//{"name":"BER_0G3"}",
      },
    },
    "{}" => {
      "pairedItems": [
        2,
      ],
      "returnData": {
        "concatenated_Warehouse": "BER_0L",
      },
    },
  },
}
`;
